libsignal-service/src/push_service.rs

```diff
@@ -147,6 +148,42 @@ impl ProfileKey {
     }
 }
 
+impl Deref for ProfileKey {
+    type Target = [u8; 32];
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl Serialize for ProfileKey {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        serializer.serialize_str(&base64::encode(self.0))
+    }
+}
+
+impl<'de> Deserialize<'de> for ProfileKey {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        Ok(ProfileKey(
+            base64::decode(String::deserialize(deserializer)?)
+                .map_err(serde::de::Error::custom)?
+                .try_into()
+                .map_err(|buf: Vec<u8>| {
+                    serde::de::Error::invalid_length(
+                        buf.len(),
+                        &"invalid profile key length",
+                    )
+                })?,
+        ))
+    }
+}
```

Comment:
I can't imagine that this isn't possible with the `with_base64` module that we already have and maybe a `#[serde(flatten)]`?

Reply:
Our `serde_base64` will not deserialize to an array of known size, but to a `Vec<u8>` so even though `Serialize` can be derived, `Deserialize` can't.

Reply:
Ah right, and for a module to do so, you probably need HKT. Dismissed!